UnitsNet/GenericMath/GenericMathExtensions.cs

```diff
@@ -2,67 +2,81 @@
-        return source.Aggregate(T.AdditiveIdentity, (acc, item) => item + acc);
+        using IEnumerator<TQuantity> e = source.GetEnumerator();
+        if (!e.MoveNext())
+        {
+            return TQuantity.AdditiveIdentity;
+        }
+
+        TQuantity result = e.Current;
+        while (e.MoveNext())
+        {
+            result += e.Current;
+        }
+
+        return result;
     }
 
     /// <summary>
-    ///     Returns the average of values.
+    ///     Calculates the arithmetic average of a sequence of vector quantities, such as Mass and Length.
     /// </summary>
+    /// <param name="source">The values.</param>
+    /// <typeparam name="TQuantity">The type of the quantity elements in the source sequence.</typeparam>
+    /// <returns>The average of the quantities, using the unit of the first element in the sequence.</returns>
+    /// <exception cref="InvalidOperationException">Thrown when the sequence is empty.</exception>
     /// <remarks>
-    ///     This method is experimental and intended to test out the new generic math interfaces implemented in .NET7 for
-    ///     UnitsNet quantities.<br />
-    ///     Generic math interfaces might replace <see cref="UnitMath" />.<br />
-    ///     Generic math LINQ support is still missing in the BCL, but is being worked on:
+    ///     Note that the generic math LINQ support is still missing in the BCL, but is being worked on:
     ///     <a href="https://github.com/dotnet/runtime/issues/64031">
     ///         API Proposal: Generic LINQ Numeric Operators · Issue
     ///         #64031 · dotnet/runtime
     ///     </a>
     /// </remarks>
-    /// <param name="source">The values.</param>
-    /// <typeparam name="T">The value type.</typeparam>
-    /// <returns>The average.</returns>
-    public static T Average<T>(this IEnumerable<T> source)
-        where T : IAdditionOperators<T, T, T>, IAdditiveIdentity<T, T>, IDivisionOperators<T, double, T>
+    public static TQuantity Average<TQuantity>(this IEnumerable<TQuantity> source)
+        where TQuantity : IQuantity, IAdditionOperators<TQuantity, TQuantity, TQuantity>, IAdditiveIdentity<TQuantity, TQuantity>,
+        IDivisionOperators<TQuantity, double, TQuantity>
     {
-        // Put accumulator on right hand side of the addition operator to construct quantities with the same unit as the values.
-        // The addition operator implementation picks the unit from the left hand side, and the additive identity (e.g. Length.Zero) is always the base unit.
-        (T value, int count) result = source.Aggregate(
-            (value: T.AdditiveIdentity, count: 0),
-            (acc, item) => (value: item + acc.value, count: acc.count + 1));
+        using IEnumerator<TQuantity> e = source.GetEnumerator();
+        if (!e.MoveNext())
+        {
+            throw new InvalidOperationException("Sequence contains no elements");
+        }
```

Comment:
Notice this: in v5 this case would have caused an exception to be thrown because of the `NaN`, however as we disabled the guard in v6- this wouldn't have thrown an exception (which I believe is the [right behavior](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.average?view=net-9.0#system-linq-enumerable-average(system-collections-generic-ienumerable((system-double))))).